src/compartment.c

```diff
@@ -239,22 +253,86 @@ void
 comp_register_ddc(struct Compartment* new_comp)
 {
     void* __capability new_ddc = cheri_address_set(manager_ddc, new_comp->base);
-    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size + new_comp->scratch_mem_stack_size);
+    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size);
     // TODO bounds double-check
     new_comp->ddc = new_ddc;
 }
 
+/* For a given Compartment `new_comp`, an address `intercept_target` pointing
+ * to a function found within the compartment which we would like to intercept,
+ * and a `intercept_data` struct representing information to perform the
+ * intercept, synthesize and inject intructions at the call point of the
+ * `intercept_target`, in order to perform a transition out of the compartment
+ * to call the appropriate function with higher privileges.
+ */
 void
 comp_add_intercept(struct Compartment* new_comp, uintptr_t intercept_target, struct func_intercept intercept_data)
 {
+    // TODO check whether negative values break anything in all these generated functions
     printf("Found `%s` func at %p.\n", intercept_data.func_name, (void*) intercept_target);
-    int offset = (intercept_data.redirect_func - intercept_target) / 4;
-    assert(offset < (1 << 27));
-    offset &= (1 << 26) - 1;
-    const int arm_b_instr_mask = 0b101 << 26;
-    uint32_t instr_binary = arm_b_instr_mask | offset;
-    int* init_addr_ptr = (void*) intercept_target;
-    struct intercept_patch new_patch = { init_addr_ptr, instr_binary };
+    int32_t new_instrs[INTERCEPT_INSTR_COUNT];
+    size_t new_instr_idx = 0;
+    uintptr_t comp_manager_cap_addr = new_comp->manager_caps + new_comp->active_manager_caps_count * sizeof(void* __capability); // TODO
+
+    const int32_t arm_function_target_register = 0b01010; // use `x10` for now
+    const int32_t arm_transition_target_register = 0b01011; // use `x11` for now
+
+    // movz x0, $target_fn_addr:lo16
+    // movk x0, $target_fn_addr:hi16
+    /*assert(intercept_target < (1 << 32));*/
+    const int32_t arm_movz_instr_mask = 0b11010010100 << 21;
+    const int32_t arm_movk_instr_mask = 0b11110010101 << 21;
```

Comment:
VIXL would save you a lot of time! It'd be easier to review, too.

Reply:
As I said, I'd be happy to help, but I honestly think that the time it would take to learn how to integrate it — not that hard really — would be quickly gained back by not having to spend time manually encoding these things. That said, it's a learning experience and perhaps useful in its own right. It should certainly be another PR.

For example, this whole `comp_add_intercept` sequence would look something like this (not tested):

```
// The boiler-plate is verbose, but you've reminded me that I wanted to propose a nicer API for this use-case.
size_t size = INTERCEPT_INSTR_COUNT * kInstructionSize;
Assembler assm(new_patch.instr, size);
assm.SetPic(PageOffsetDependentCode);  // Otherwise 'adrp' will complain.
{  // Nested scope to ensure that the guard destructs before the assembler.
  CodeBufferCheckScope guard(
      &assm, size, CodeBufferCheckScope::kDontReserveBufferSpace, CodeBufferCheckScope::kExactSize);

  // We typically `#define __ assm.` somewhere.
  __ movz(x0, intercept_data.redirect_func & 0xffff);
  __ movk(x0, (intercept_data.redirect_func >> 16) & 0xffff);
  __ adrp(x11, (comp_manager_cap_addr >> 12) - (intercept_target >> 12));
  __ ldr(c11, MemOperand(x11, align_down(comp_manager_cap_addr - intercept_target, 1 << 12)));
  ptraddr_t arm_b_instr_offset = ...;
  __ b(arm_b_instr_offset);
}
assm.FinalizeCode();
CPU::EnsureIAndDCacheCoherency(new_patch.instr, size);
```

One caveat is that it doesn't build for purecap yet. I've plans to fix that, but haven't found the time.